File: harvard/CS50x/W06/credit/credit.py
from cs50 import get_string


def is_valid(number_str):
    """Checks if a credit card number is valid using Luhn's algorithm."""
    n = len(number_str)
    if not number_str.isdigit():
        return False

    sum_val = 0
    for i in range(n - 1, -1, -1):
        digit = int(number_str[i])
        if (n - 1 - i) % 2 == 1:
            digit *= 2
            if digit > 9:
                digit -= 9
        sum_val += digit
    return (sum_val % 10) == 0


def main():
    number_str = get_string("Number: ")

    if not is_valid(number_str):
        print("INVALID")
        return

    n = len(number_str)
    first_two = number_str[:2]
    first_digit = number_str[0]

    if (n == 15) and (first_two == "34" or first_two == "37"):
        print("AMEX")
    elif (n == 16) and (first_two in ["51", "52", "53", "54", "55"]):
        print("MASTERCARD")
    elif (n >= 13 and n <= 16) and (first_digit == "4"):
        print("VISA")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()

File: harvard/CS50x/W01/credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

bool is_valid(long number);
string get_card_type(long number);

int main(void)
{
    long number;
    do
    {
        number = get_long("Number: ");
    }
    while (number < 1);

    if (is_valid(number))
    {
        printf("%s\n", get_card_type(number));
    }
    else
    {
        printf("INVALID\n");
    }
}

bool is_valid(long number)
{
    int sum = 0;
    bool second = false;

    while (number > 0)
    {
        int digit = number % 10;

        if (second)
        {
            digit *= 2;
            if (digit > 9)
                digit = digit / 10 + digit % 10;
        }

        sum += digit;
        number /= 10;
        second = !second;
    }

    return (sum % 10 == 0);
}

string get_card_type(long number)
{
    int length = 0;
    long start = number;

    while (start >= 100)
    {
        start /= 10;
        length++;
    }
    length += 2;

    int first_digit = start / 10;
    int first_two = start;

    if ((length == 15) && (first_two == 34 || first_two == 37))
        return "AMEX";
    else if ((length == 16) && (first_two >= 51 && first_two <= 55))
        return "MASTERCARD";
    else if ((length == 13 || length == 16) && (first_digit == 4))
        return "VISA";
    else
        return "INVALID";
}
